File: app/src/main/java/ru/melodin/fast/adapter/PopupAdapter.java
package ru.melodin.fast.adapter;

import android.content.Context;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;

import java.util.ArrayList;

import ru.melodin.fast.R;
import ru.melodin.fast.model.ListItem;

public class PopupAdapter extends RecyclerAdapter<ListItem, PopupAdapter.ViewHolder> {

    public static final int ID_CLEAR_DIALOG = 0;
    public static final int ID_NOTIFICATIONS = 1;
    public static final int ID_LEAVE = 2;

    public PopupAdapter(Context context, ArrayList<ListItem> values) {
        super(context, values);
    }

    public int searchPosition(int id) {
        for (ListItem item : getValues())
            if (item.getId() == id)
                return getValues().indexOf(item);

        return -1;
    }

    public void updateItem(@NonNull ListItem item) {
        for (ListItem i : getValues())
            if (i.getId() == item.getId()) {
                i.setTitle(item.getTitle());
                i.setIcon(item.getIcon());
                i.setVisible(item.isVisible());
                break;
            }
    }

    @Override
    public void changeItems(ArrayList<ListItem> items) {
        super.changeItems(items);
        notifyDataSetChanged();
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new ViewHolder(inflater.inflate(R.layout.activity_messages_popup_item, parent, false));
    }

    class ViewHolder extends RecyclerHolder {

        ImageView icon;
        TextView title;

        ViewHolder(@NonNull View v) {
            super(v);

            title = v.findViewById(R.id.title);
            icon = v.findViewById(R.id.icon);
        }

        @Override
        public void bind(int position) {
            ListItem item = getItem(position);

            icon.setImageDrawable(item.getIcon());
            title.setText(item.getTitle());
        }
    }
}

File: app/src/main/java/ru/melodin/fast/adapter/PopupAdapter.kt
package ru.melodin.fast.adapter

import android.content.Context
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import ru.melodin.fast.R
import ru.melodin.fast.model.ListItem
import java.util.*

class PopupAdapter(context: Context, values: ArrayList<ListItem>) :
    RecyclerAdapter<ListItem, PopupAdapter.ViewHolder>(
        context,
        R.layout.activity_messages_popup_item,
        values
    ) {

    override fun changeItems(items: ArrayList<ListItem>) {
        super.changeItems(items)
        notifyDataSetChanged()
    }

    override fun getItemViewType(position: Int): Int {
        return getItem(position).type
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerHolder {
        return when (viewType) {
            ListItem.TYPE_ITEM -> ViewHolder(getView(parent)!!)
            ListItem.TYPE_DIVIDER -> Holder(inflater.inflate(R.layout.view_divider, parent, false))
            else -> Holder(View(context))
        }
    }

    inner class ViewHolder(v: View) : RecyclerHolder(v) {

        var icon: ImageView = v.findViewById(R.id.icon)
        var title: TextView = v.findViewById(R.id.title)

        override fun bind(position: Int) {
            val item = getItem(position)

            icon.setImageDrawable(item.icon)
            title.text = item.title
        }
    }

    inner class Holder(v: View) : RecyclerHolder(v)

    companion object {
        const val ID_CLEAR_HISTORY = 0
        const val ID_NOTIFICATIONS = 1
        const val ID_LEAVE = 2
        const val ID_CHAT_INFO = 3
    }
}